src/libraries/Microsoft.PowerFx.Core/Types/DefinedTypeResolver.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.IR;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Syntax.Visitors;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Types
+{
+    internal class DefinedTypeResolver
+    {
+        private readonly Dictionary<string, DefinedType> _typesDict;
+        private readonly ReadOnlySymbolTable _globalSymbols;
+
+        private readonly IEnumerable<string> _nodes;
+        private readonly List<TopologicalSortEdge<string>> _edges;
+
+        private readonly List<TexlError> _errors;
+
+        // TODO: Add more future type names
```

Comment:
>Add more future type names [](http://example.com/codeflow?start=17&length=26)

breaking to add ?

Reply:
@gregli-msft can you suggest the names that we should be restricting from being created ? will resolve this Todo in this PR

Reply:
will fix as we get more input from greg

Reply:
Are those names already restricted? If not, we should probably add those: ThisItem, ThisRecord, ThisGroup, Self